File: backend/app/schemas/portfolio.py
"""Portfolio-level aggregation schemas for dashboard views."""

from pydantic import Field

from app.schemas.base import CamelCaseModel


class SeverityCount(CamelCaseModel):
    """A single severity bucket with its count."""

    severity: str
    count: int = Field(ge=0, default=0)


class HeatMapEntry(CamelCaseModel):
    """Single cell in the repos x severity heat map."""

    project_id: str
    project_name: str
    critical: int = Field(ge=0, default=0)
    high: int = Field(ge=0, default=0)
    medium: int = Field(ge=0, default=0)
    low: int = Field(ge=0, default=0)
    info: int = Field(ge=0, default=0)


class TopRepo(CamelCaseModel):
    """A repo ranked by risk (total weighted findings)."""

    project_id: str
    project_name: str
    total_findings: int = Field(ge=0)
    critical_count: int = Field(ge=0, default=0)
    high_count: int = Field(ge=0, default=0)
    risk_score: float = Field(ge=0, description="Weighted risk score")
    provider: str = ""
    quantum_risk: float = Field(ge=0, default=0, description="% quantum-vulnerable findings")
    compliance_percent: float = Field(ge=0, default=0, description="Avg compliance score")
    last_scan: str | None = None


class PortfolioSummary(CamelCaseModel):
    """Aggregate portfolio summary across all accessible repos."""

    total_repos: int = Field(ge=0)
    total_findings: int = Field(ge=0)
    critical_plus_high: int = Field(ge=0, default=0)
    affected_repos: int = Field(ge=0, default=0, description="Repos with critical or high findings")
    quantum_exposed: int = Field(ge=0, default=0, description="Count of quantum-vulnerable findings")
    quantum_exposed_percent: float = Field(ge=0, le=100, default=0)
    quantum_readiness_percent: float = Field(
        ge=0, le=100, default=100, description="Percentage of findings that are quantum-safe"
    )
    compliance_avg: float = Field(ge=0, le=100, default=0)
    compliance_framework: str = ""
    last_scan_time: str | None = None
    severity_distribution: list[SeverityCount] = Field(default_factory=list)
    heat_map: list[HeatMapEntry] = Field(default_factory=list)
    top_risk_repos: list[TopRepo] = Field(default_factory=list)


class FrameworkScore(CamelCaseModel):
    """Compliance score for a single framework across all repos."""

    framework: str
    avg_score: float = Field(ge=0, le=100)
    total_compliant: int = Field(ge=0)
    total_non_compliant: int = Field(ge=0)
    repos_evaluated: int = Field(ge=0)


class PortfolioCompliance(CamelCaseModel):
    """Per-framework compliance scores across all repos."""

    frameworks: list[FrameworkScore]


class QuantumCategory(CamelCaseModel):
    """Counts for a quantum readiness category."""

    vulnerable: int = Field(ge=0, default=0)
    safe: int = Field(ge=0, default=0)
    broken: int = Field(ge=0, default=0)
    unknown: int = Field(ge=0, default=0)


class MigrationPriorityItem(CamelCaseModel):
    """An algorithm needing migration, aggregated across all repos."""

    algorithm: str
    total_count: int = Field(ge=0)
    affected_repos: int = Field(ge=0)
    migrate_to: str = Field(default="")
    priority: int = Field(ge=1)


class PortfolioQuantum(CamelCaseModel):
    """Aggregate quantum readiness across all accessible repos."""

    counts: QuantumCategory
    migration_priority: list[MigrationPriorityItem]

File: frontend/src/mocks/data/portfolio.ts
/**
 * Mock data for the Portfolio Dashboard — heat map, risk scores, top repos.
 */

export type SeverityLevel = 'critical' | 'high' | 'medium' | 'low' | 'info';

export interface HeatMapCell {
  repoId: string;
  repoName: string;
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
}

export interface TopRiskyRepo {
  repoId: string;
  repoName: string;
  provider: string;
  findings: number;
  critical: number;
  quantumRisk: number;
  compliancePercent: number;
  lastScan: string;
}

export interface SeverityDistribution {
  severity: SeverityLevel;
  count: number;
}

export interface PortfolioSummaryData {
  totalFindings: number;
  totalRepos: number;
  criticalPlusHigh: number;
  affectedRepos: number;
  quantumExposed: number;
  quantumExposedPercent: number;
  complianceAvg: number;
  complianceFramework: string;
  quantumReadinessPercent: number;
  lastScanTime: string;
  severityDistribution: SeverityDistribution[];
  topRiskRepos: TopRiskyRepo[];
}

export interface HeatMapGroup {
  id?: string;
  name: string;
  projectCount?: number;
  criticalCount?: number;
  highCount?: number;
  mediumCount?: number;
  lowCount?: number;
  infoCount?: number;
}

export interface HeatMapData {
  repos?: HeatMapCell[];
  groups?: HeatMapGroup[];
}

export const MOCK_PORTFOLIO_SUMMARY: PortfolioSummaryData = {
  totalFindings: 1247,
  totalRepos: 8,
  criticalPlusHigh: 89,
  affectedRepos: 5,
  quantumExposed: 312,
  quantumExposedPercent: 42,
  complianceAvg: 72,
  complianceFramework: 'NIST 800-131A',
  quantumReadinessPercent: 58,
  lastScanTime: '2min ago',
  severityDistribution: [
    { severity: 'critical', count: 12 },
    { severity: 'high', count: 77 },
    { severity: 'medium', count: 234 },
    { severity: 'low', count: 156 },
    { severity: 'info', count: 768 },
  ],
  topRiskRepos: [
    {
      repoId: 'repo-4',
      repoName: 'mobile-backend',
      provider: 'Bitbucket',
      findings: 456,
      critical: 3,
      quantumRisk: 68,
      compliancePercent: 45,
      lastScan: '1d ago',
    },
    {
      repoId: 'repo-1',
      repoName: 'payment-service',
      provider: 'GitHub',
      findings: 342,
      critical: 5,
      quantumRisk: 72,
      compliancePercent: 65,
      lastScan: '2h ago',
    },
    {
      repoId: 'repo-2',
      repoName: 'auth-api',
      provider: 'GitHub',
      findings: 187,
      critical: 2,
      quantumRisk: 54,
      compliancePercent: 82,
      lastScan: '3h ago',
    },
    {
      repoId: 'repo-3',
      repoName: 'data-pipeline',
      provider: 'GitLab',
      findings: 98,
      critical: 0,
      quantumRisk: 22,
      compliancePercent: 94,
      lastScan: '5h ago',
    },
    {
      repoId: 'repo-5',
      repoName: 'api-gateway',
      provider: 'GitHub',
      findings: 64,
      critical: 1,
      quantumRisk: 38,
      compliancePercent: 78,
      lastScan: '6h ago',
    },
    {
      repoId: 'repo-6',
      repoName: 'user-service',
      provider: 'GitHub',
      findings: 42,
      critical: 0,
      quantumRisk: 30,
      compliancePercent: 85,
      lastScan: '8h ago',
    },
    {
      repoId: 'repo-7',
      repoName: 'notification-svc',
      provider: 'GitLab',
      findings: 34,
      critical: 0,
      quantumRisk: 18,
      compliancePercent: 91,
      lastScan: '12h ago',
    },
    {
      repoId: 'repo-8',
      repoName: 'config-service',
      provider: 'GitHub',
      findings: 24,
      critical: 0,
      quantumRisk: 12,
      compliancePercent: 96,
      lastScan: '1d ago',
    },
  ],
};

export const MOCK_HEAT_MAP: HeatMapData = {
  repos: [
    { repoId: 'repo-1', repoName: 'payment-service', critical: 5, high: 28, medium: 67, low: 42, info: 200 },
    { repoId: 'repo-2', repoName: 'auth-api', critical: 2, high: 12, medium: 38, low: 35, info: 100 },
    { repoId: 'repo-3', repoName: 'data-pipeline', critical: 0, high: 4, medium: 18, low: 26, info: 50 },
    { repoId: 'repo-4', repoName: 'mobile-backend', critical: 3, high: 22, medium: 68, low: 33, info: 330 },
    { repoId: 'repo-5', repoName: 'api-gateway', critical: 1, high: 6, medium: 22, low: 12, info: 23 },
    { repoId: 'repo-6', repoName: 'user-service', critical: 0, high: 3, medium: 12, low: 8, info: 19 },
    { repoId: 'repo-7', repoName: 'notification-svc', critical: 0, high: 2, medium: 8, low: 6, info: 18 },
    { repoId: 'repo-8', repoName: 'config-service', critical: 1, high: 0, medium: 1, low: 4, info: 18 },
  ],
};

export function getPortfolioSummary(): PortfolioSummaryData {
  return MOCK_PORTFOLIO_SUMMARY;
}

export function getHeatMap(): HeatMapData {
  return MOCK_HEAT_MAP;
}
